From 622727269787ad742891526b427d98ee0ce27c32 Mon Sep 17 00:00:00 2001
From: Steve deRosier <steve.derosier@lairdtech.com>
Date: Wed, 29 Jun 2016 18:03:40 -0700
Subject: [PATCH] laird: update version to laird15

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0d63c5d..83cfbad 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 4
 PATCHLEVEL = 1
 SUBLEVEL = 13
-EXTRAVERSION = -laird14
+EXTRAVERSION = -laird15
 NAME = Series 4800
 
 # *DOCUMENTATION*
-- 
1.9.1

